README.md
>_Forked from https://github.com/mikebanks/AbuseIPdbSCAN_

# AbuseIP DB Scanner

This is a **python3** script that will parse any text file containg IP addresses and return relevant threat information using the AbuseIPDB API.

# New Releases

Visit the **[Latest Releases](https://github.com/louigigr/AbuseIP-db-scanner/releases/latest)** page to download the [python version](https://github.com/louigigr/AbuseIP-db-scanner/releases/download/v1.0-stable/abuseipdb-generic.zip) or the [windows binary version](https://github.com/louigigr/AbuseIP-db-scanner/releases/download/v1.0-stable/abuseipdb-win-v1.0.zip) of the script.

## Download and use binary

- Download the [windows binary](https://github.com/louigigr/AbuseIP-db-scanner/releases/download/v1.0-stable/abuseipdb-win-v1.0.zip) from the [releases](https://github.com/louigigr/AbuseIP-db-scanner/releases) page

>![alt text](https://cdn.els-cdn.com/sd/img/sprite_parts/warning_a.gif "Important!") **Important! The binary should work fine on Windows 10, in any other case you need to download  [Windows 10 Universal C Runtime](https://www.microsoft.com/en-us/download/details.aspx?id=48234)**![alt text](https://cdn.els-cdn.com/sd/img/sprite_parts/warning_a.gif "Important!")

- Unzip the downloaded file to a folder of your choice _(e.g C:\users\user\programs\)_
- Open windows terminal _(windows key + R and type CMD)_
- Change to the desired directory _(cd \users\user\programs\)_

> ![alt text](http://www.happystove.com/images/starYellow.png "Optional") Optional: _You can add the directory to your PATH environment variable or copy the script inside C:\Windows\System32\ directory in order for it to be available to you at any time while on the terminal._![alt text](http://www.happystove.com/images/starYellow.png "Optional")
- Run the script with ``` abuseipdb  ```
- You will be prompted to enter your api key ``` Please enter your abuseipdb.com API key here: ``` _1234567_
- The key is stored in registry and you can change it at any time using the **-a** switch
- To properly use the program you need to have a text file _(e.g alerts.log)_ that contains IP addresses in any form. The file is parsed and all non-local IP addresses are extracted and deduplicated. In order to execute the script just type: 
``` abuseipdb -f {FileWithIPs} ```
e.g. ``` abuseipdb -f alert.log ```

> Optional: To save to a file you can apend > {filename} after the command e.g ``` abuseipdb -f alert.log > ThreatReport.txt ```

- Specify the number of days you wish to search and display alerts for by using the -d or --days switch.
e.g. ``` abuseipdb -f myip.txt -d 60 ```

  - Accepted range is 1 to 365
  - If no switch is specified the script uses a default value of _30_
  - Contains a really simple failsafe so the script won't break, really helps with the binary version
- To get help type abuseipdb -h

## Setup and Usage with python

### Setup

- Python3 (_2.7 may have errors_)
- Requests
```
pip3 install Requests
```
- Requests[security]
```
pip3 install requests[security]
```

### Usage:

In order to use the script you will need an API key. The API key is stored in a file _my.api_ in the same directory as the script. In order for the script to work make sure you edit the file _my.api_ and enter your API Key. API key information can be found here: (https://www.abuseipdb.com/api.html)

To use type:

```
python3 abuseipdb.py -f file_to_parse.txt
```

 The options are as follows:

```
-t      outputs items in tab seperated values (Default)

-c      outputs items in comma seperated values

-d      specifies number of days
```


## Setup and use with PIPENV

### Setup PIPENV and AbuseIP-db-scanner

* Requirements
  * Python3
  * Pipenv _(install with pip install pipenv)_
  * Requests

```
cd ~
git clone https://github.com/louigigr/AbuseIP-db-scanner.git
cd AbuseIP-db-scanner
pipenv install --python 3.6
pipenv shell
pipenv install requests
```

### Usage

To use type:

```
python abuseipdb.py
```

## Troubleshooting

- If you are receiving errors, please look at the Issues queue and see if there is already an issue open.

- If you have a unique issue, please create a new Issue, and include the output of your terminal from the bootstrap script down until the error.



releases.md
## Release v1.0
First stable release of AbuseIP-db-scanner 

> Windows binaries and Python3 version available

## Download

> [Python version](https://github.com/louigigr/AbuseIP-db-scanner/releases/download/v1.0-stable/abuseipdb-generic.zip)

> [Windows binary version](https://github.com/louigigr/AbuseIP-db-scanner/releases/download/v1.0-stable/abuseipdb-win-v1.0.zip)

_You can also visit the [Repository Releases page ](https://github.com/louigigr/AbuseIP-db-scanner/releases)_

**_The windows binary should work fine in Windows 10. In any other case you may need to download [Windows 10 Universal C Runtime](https://www.microsoft.com/en-us/download/details.aspx?id=48234)_**

## What's New!
* Save API Key to a file
* Added -d or --days switch to keep generic script as a complete CLI solution which might be better for automation
* Retrieve Reports by specifying number of days. If no number of days is specified the default value of _30_ is used


### Features

* Registry API key Storage (Windows only)
* Input Key on first run
* Store API Key to file (python version)
* Export to CSV
* Export to Tab Seperated Values
* IP parser that will automatically:
  * Grab all IP addresses in the file no matter the formatting
  * Detect and remove duplicate IP addresses
  * Detect and remove local subnet IP addresses _(192.168.0.0/16, 172.16.0.0/12, 10.0.0.0/8)_



## Generic Release

The generic release uses the file method to store your API key.


abuseipdb.py
#! /bin/python/env
import os
import codecs
import argparse
import re
import sys
import requests
import ipaddress
import socket
import time
from sys import argv
import winreg

REG_PATH = r"Software\QgSoft"

def set_reg(name, value):
	try:
		winreg.CreateKey(winreg.HKEY_CURRENT_USER, REG_PATH)
		registry_key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, REG_PATH, 0, 
									   winreg.KEY_WRITE)
		winreg.SetValueEx(registry_key, name, 0, winreg.REG_SZ, value)
		winreg.CloseKey(registry_key)
		return True
	except WindowsError:
		return False

def get_reg(name):
	try:
		registry_key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, REG_PATH, 0,
									   winreg.KEY_READ)
		value, regtype = winreg.QueryValueEx(registry_key, name)
		winreg.CloseKey(registry_key)
		return value
	except WindowsError:
		return None

if not get_reg('apikey'):
	apikey=input("Please enter your abuseipdb.com API key here: ")
	set_reg('apikey', str(apikey))
	print("API Key stored\n")
	print("Reading from registry: " + str(get_reg('apikey')))
else:
	print("API Key found stored in registry.\nUsing API KEY: " + str(get_reg('apikey')) + "\n")
api_key = str(get_reg('apikey'))

parser = argparse.ArgumentParser(
	formatter_class=argparse.RawDescriptionHelpFormatter,
	description='Part of the Quadrant Global Security Toolset.\n\nThis program uses the AbuseIPDB.com API in order to perform queries and detect malicious IP addresses.\nThe input can be any text file in any form, the IP addresses are parsed and recognised automaticaly.\n'
)
required = parser.add_argument_group('required arguments')
required.add_argument(
	"-f",
	"--file",
	help="parses IP Addresses from a single given file",
	action="store",
	required=False)
required.add_argument(
	"-d",
	"--days",
	help="Number of days to look back in history for alerts",
	action="store",
	required=False)


parser.add_argument("-t", "--tsv", help="outputs items in tab seperated values (Default)", action="store_true")
parser.add_argument("-c", "--csv", help="outputs items in comma seperated values",  action="store_true")
parser.add_argument("-a", dest="APIKEY", help="stores new API key in registry",  action="store")
#parser.add_argument("-d", dest="DAYS", help="Number of days to look back in history for alerts",  action="store")

args = parser.parse_args()



def get_file(infile):
	with codecs.open(infile, "r", encoding='utf-8', errors='ignore') as f:
		return f.read()


def get_cat(x):
	return {
		3: 'Frad_Orders',
		4: 'DDoS_Attack',
		5: 'FTP_Brute-Force',
		6: 'Ping of Death',
		7: 'Phishing',
		8: 'Fraud VoIP',
		9: 'Open_Proxy',
		10: 'Web_Spam',
		11: 'Email_Spam',
		12: 'Blog_Spam',
		13: 'VPN IP',
		14: 'Port_Scan',
		15: 'Hacking',
		16: 'SQL Injection',
		17: 'Spoofing',
		18: 'Brute_Force',
		19: 'Bad_Web_Bot',
		20: 'Exploited_Host',
		21: 'Web_App_Attack',
		22: 'SSH',
		23: 'IoT_Targeted',
	}.get(
		x,
		'UNK CAT, ***REPORT TO MAINTAINER***OPEN AN ISSUE ON GITHUB w/ IP***')


def get_report(IP):
	if not args.days:
		print("Setting days to default: 30")
		noofdays = 30
	else:
		noofdays=args.days
	try:
		val=int(noofdays)
	except ValueError:
		print("\nYou must specify an integer between 1 and 365.\n")
		sys.exit()
	if (int(noofdays) <= int(0) or int(noofdays) > int(365)):
		print("\nYou must specify an integer between 1 and 365.\n")
		sys.exit()

	#request = 'https://www.abuseipdb.com/check/%s/json?key=%s&days=%s' % (IP, api_key, str(argv[4]))
	request = 'https://www.abuseipdb.com/check/%s/json?key=%s&days=%s' % (IP, api_key, noofdays)
	# DEBUG
	#print(request)
	r = requests.get(request)
	# DEBUG
	#print(r.json())
	try:
		data = r.json()
		if data == []:
			print("%s:  No Abuse Reports" % IP)
		else:
			for record in data:
				log = []
				ip_address = ("Alert for %s:" % IP)
				# ip_address = record['ip']
				country = record['country']
				# iso_code = record['isoCode']
				category = record['category']
				created = record['created']
				log.append(ip_address)
				log.append(country)
				# log.append(iso_code)
				log.append(created)
				for cat in category:
					temp_cat = get_cat(cat)
					log.append(temp_cat)
					if args.csv:
						print(','.join(log))
					else: 
						print('\t'.join(log))
					log.remove(temp_cat)
	except (ValueError, KeyError, TypeError):
		#log = []
		ip_address = ("Alert for %s:" % IP)
		# ip_address = record['ip']
		country = data['country']
		# iso_code = record['isoCode']
		category = data['category']
		created = data['created']
		log.append(ip_address)
		log.append(country)
		# log.append(iso_code)
		log.append(created)
		for cat in category:
			temp_cat = get_cat(cat)
			log.append(temp_cat)
			if args.csv:
				print(','.join(log))
			else: 
				print('\t'.join(log))
			log.remove(temp_cat)

def Deduplicate(duplicate):
	final_list = []
	for num in duplicate:
		if num not in final_list:
			final_list.append(num)
	return final_list


def main():

	# if not args.days:
		# print("You must specify number of days for the report to run\nE.g: abuseipdb -f myfile.txt -d 365")
		# sys.exit()

	if args.APIKEY:
		print("Reset API REQUEST")
		set_reg('apikey', str(argv[2]))
		print("\nYour key is: " + str(get_reg('apikey')))
		sys.exit()
	if args.file:
		f = get_file(argv[2])
		found = Deduplicate(re.findall(
			r'(?:[\d]{1,3})\.(?:[\d]{1,3})\.(?:[\d]{1,3})\.(?:[\d]{1,3})', f))

		list(set(found))

		print("Abuse IP Database REPORT:")
		count = 0
		for ip in found:
			try:
				socket.inet_aton(ip)
			except socket.error:
				continue

			if ipaddress.ip_address(ip).is_private is False:
				if count == 59:
					time.sleep(60)
					count = 0
				get_report(ip)
				count += 1


if __name__ == '__main__':
	main()
